drt - Variable stop durations

nkuehnel opened the pull request:
New attempt of 
https://github.com/matsim-org/matsim-libs/pull/1809

I tried to incorporate the ideas discussed in the old PR and for now chose "IncrementalStopDurationEstimator" over "AdditionalStopDuration" for naming the second interface.

@sebhoerl @michalmac let me know if this is what you envisioned.

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:
```diff
@@ -10,94 +10,97 @@
 		if (dropoff.newWaypoint.getLink() == dropoff.previousWaypoint.getLink()) {
 			// no detour, no additional stop duration
-			double dropoffTimeLoss = 0;
+			double dropoffTimeLoss = calcAdditionalDropoffStopDurationIfSameLinkAsPrevious(vEntry, insertion.dropoff.index, drtRequest);
```
please update the comment in L 83

nkuehnel replied:
✅

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:
```diff
@@ -109,18 +112,44 @@ private PickupDetourInfo calcPickupDetourInfoIfPickupToDropoffDetour(VehicleEntr
+			if (startTask.isPresent() && STOP.isBaseTypeOf(startTask.get())) {
+				stopTask = (DrtStopTask) startTask.get();
+			} else {
+				return stopDurationEstimator.calcForPickup(vEntry.vehicle, null, drtRequest);
```
replace with `stopTask = null;` ?

michalmac replied:
L130 ^

nkuehnel replied:
✅

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/InsertionDetourTimeCalculator.java:
```diff
@@ -109,18 +112,44 @@ private PickupDetourInfo calcPickupDetourInfoIfPickupToDropoffDetour(VehicleEntr
+			if (startTask.isPresent() && STOP.isBaseTypeOf(startTask.get())) {
+				stopTask = (DrtStopTask) startTask.get();
+			} else {
+				return stopDurationEstimator.calcForDropoff(vEntry.vehicle, null, drtRequest);
```
ditto

nkuehnel replied:
✅

michalmac commented on contribs/drt/src/main/java/org/matsim/contrib/drt/schedule/DrtStayTaskEndTimeCalculator.java:
```diff
@@ -30,10 +30,10 @@
 public class DrtStayTaskEndTimeCalculator implements ScheduleTimingUpdater.StayTaskEndTimeCalculator {
 
-	final double stopDuration;
+	private StopDurationEstimator stopDurationEstimator;
```
`final`?

nkuehnel replied:
✅

michalmac reviewed (approved):
Awesome work @nkuehnel and @sebhoerl! 🚀

michalmac commented:
@nkuehnel I think at one point you did some rebasing or cherry picking (instead of merging master into this branch), so my commits got duplicated. Maybe rebase this branch on origin/master. They should simply disappear (as there haven't been a lot of changes recently in drt).

nkuehnel commented on contribs/drt/src/main/java/org/matsim/contrib/drt/optimizer/insertion/DefaultIncrementalStopDurationEstimator.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class DefaultIncrementalStopDurationEstimator implements IncrementalStopDurationEstimator {
+
+	private final double staticStopDuration;
```
✅

nkuehnel commented:
@michalmac thanks for your feedback, your suggestions should be factored in now. Also I rebased branches locally and updated the remote branch accordingly.

sebhoerl commented:
Thanks :)
